source/extensions/certificate_providers/certificate_provider.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace CertificateProviders {
+
+struct Certpair {
+  const std::string& certificate_;
+  const std::string& private_key_;
+};
+
+struct Capabilites {
+  /* whether or not a provider provides a ca cert directly */
+  bool provide_ca_cert = false;
+
+  /* Whether or not a provider provides identity certs directly */
+  bool provide_identity_certs = false;
+
+  /* whether or not a provider supports generating identity certs during handshake,
+   * which requires the capability provide_ca_cert to sign the certificates
+   */
+  bool generate_identity_certs = false;
+};
+
+class CertificateProvider {
+public:
+  virtual ~CertificateProvider() = default;
+
+  virtual Capabilites capabilities() const PURE;
+
+  /**
+   * @return CA certificate from provider
+   */
+  virtual const std::string& getCACertificate() const PURE;
```

Comment:
This accessor-style interface is sufficient for a static provider impl, or even for a polling-based provider (i.e., one that explicitly goes out and fetches the cert somehow each time it is accessed).  However, consider the case of a watch-based provider, where the impl will start a watch on some remote service for a particular cert and will receive updates whenever that cert changes.  In order to support that, I think the cert provider will need to know which certificate names to subscribe to *before* the cert is actually needed -- in other words, we will want to subscribe to a cert name as soon as we get a config that uses that name, and we will want to unsubscribe from that cert name when the config changes to no longer use that name.

In gRPC, we have a watch-based interface for this.  We have a class called [`grpc_tls_certificate_distributor`](https://github.com/grpc/grpc/blob/7c74d58cbecc8342c349cbf39eecb9aa0b666522/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L37) that is used by provider implementations to distribute certificates to watchers.  The code that wants to subscribe to a certificate will implement [`TlsCertificateWatcherInterface`](https://github.com/grpc/grpc/blob/7c74d58cbecc8342c349cbf39eecb9aa0b666522/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L41), and it will pass an instance of that class to [`WatchTlsCertificates()`](https://github.com/grpc/grpc/blob/7c74d58cbecc8342c349cbf39eecb9aa0b666522/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L144) to start a watch on a particular cert name.  The watch can then be cancelled by calling [`CancelTlsCertificatesWatch()`](https://github.com/grpc/grpc/blob/7c74d58cbecc8342c349cbf39eecb9aa0b666522/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L152).  Meanwhile, the cert provider impl registers a callback with the distributor via [`SetWatchStatusCallback()`](https://github.com/grpc/grpc/blob/7c74d58cbecc8342c349cbf39eecb9aa0b666522/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L127), so that it knows what certs need to be watched.  When it has new cert data available, it will set it via [`SetKeyMaterials()`](https://github.com/grpc/grpc/blob/7c74d58cbecc8342c349cbf39eecb9aa0b666522/src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.h#L81).  Whenever the distributor receives new certs from the provider, it will automatically notify all watchers of that cert name.

I think you will probably want a similar style of interface here.

Reply:
I don't fully understand this. Let me spend some time to digest it.

Thanks for your comments

Reply:
@markdroth 
I looked the gRPC certificate provider and envoy xDS, they both have the "watcher" or "subscription" mechanism.

In envoy, subscriptions is used for updating xDS configs, there are several implementations including [FilesystemSubscriptionImpl](https://github.com/envoyproxy/envoy/blob/main/source/common/config/filesystem_subscription_impl.cc), [GrpcSubscriptionImpl](https://github.com/envoyproxy/envoy/blob/main/source/common/config/grpc_subscription_impl.cc), etc. They all have [ConfigSource](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/config_source.proto#config-core-v3-configsource) as one of inputs for construction.

I'm not sure if you want the similar mechanism like xDS update. But I think what you mentioned is to watch concrete certificate update, that's a little weird in Envoy, because Envoy watches whole update for LDS or CDS, etc.

Refer  to listener implementation, it can be configured in [StaticResources](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto#config-bootstrap-v3-bootstrap-staticresources) or [DynamicResources](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto#config-bootstrap-v3-bootstrap-staticresources). So I'm thinking about maybe we need to separate the certificate providers to static config and dynamic condig as well, currently the certificate providers API looks like static resources. If we use it in dynamical way, I think we should limit it to [ConfigSource](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/config_source.proto#config-core-v3-configsource)

Reply:
I think the key requirement here is the need to support the case where the cert name comes from the [`CertificateProviderPluginInstance.certificate_name` field](https://github.com/envoyproxy/envoy/blob/7900682f09b25163445e83042fe7068ba0bbbd32/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L269) instead of being part of the definition of the provider plugin in the bootstrap file.

I don't think it actually matters what specific mechanism is used to watch the certs, but let's take your example of a hypothetical DPDS resource type.  Let's say that there is one DPDS resource for each cert, and the client subscribes to each resource by name based on the name of the cert it needs.

In this model, when does the provider send the xDS subscription request for the DPDS resource for a given cert?  This API does not provide it any way to know the cert name prior to the point at which the cert is actually needed (e.g., to handle an incoming connection), which means that it would need to send the xDS subscription request at that point, which delays handling of incoming connections.  And even worse, the API here requires the provider to return the cert synchronously, which means you'd have to block a thread until the xDS server returns the requested resource (which is incompatible with Envoy's threading model).

And even if we did have the provider kick off the xDS request at the point at which the cert is needed, how does it know how long to keep the xDS subscription active for?  We don't want it to simply throw it away and unsubscribe as soon as it returns the result, because there may be multiple connections coming in that will all want to use the same cert, and it would be extremely inefficient to have to fetch it from the xDS server each time a connection comes in.  But we also don't want to cache it forever, because if we get a config change that tells us to stop using that cert, it would be wasteful to keep it cached (and keep receiveing xDS updates for it) when we don't need it anymore.

So I think what we need here is an explicit way to tell the provider that we want to subscribe to a cert when we first receive a config that includes a `CertificateProviderPluginInstance` for a particular cert name, and a way to unsubscribe to the cert when we get a config update that removes that `CertificateProviderPluginInstance`.

In between those two points, I suppose it doesn't really matter whether the cert itself is delivered to the caller whenever new data is available or whether the caller explicitly requests it from the cert provider each time it needs it.  However, there are two potential complications with having the caller explicitly request the cert each time it needs it:

1. There may be synchronization issues that need to be taken into account in case the resource is being updated by one thread at the same time as callers in other threads are trying to access it.  I don't know how Envoy typically handles this.
2. If the caller tries to access the cert after the client has subscribed to the xDS resource but before the xDS server has sent the resource to it, then the cert might not yet be available, which means that the API needs to be able to handle getting an async result.

I think that since the caller needs to explicitly start and stop the watch on the provider anyway, it probably makes sense to just use a similar pattern to gRPC here, where the caller registers a watcher object that will be notified of updates whenever the cert changes.

I would also note that this pattern seems consistent with the long-term direction that we've discussed for xDS subscriptions in general in #13009.

Reply:
I think there might be some misunderstanding. 

> If the caller tries to access the cert after the client has subscribed to the xDS resource but before the xDS server has sent the resource to it, then the cert might not yet be available, which means that the API needs to be able to handle getting an async result.

I don't think this API needs to handling this issue, it's should be responsibility of the caller. For instance, we have unimplemented API , [tls_certificate_provider_instance](https://github.com/envoyproxy/envoy/blob/9d5627a0879b0a029e90515137c108e1d2884bfc/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L247), we might need utilize SSL_CTX_set_select_certificate_cb to select certificate from certificate provider, if we don't find the cert for handshake, we rely on SSL asynchronous mechmism to retry later.

Reply:
I think we need open new conversation on latest code, this code and my words are out of date.